src/hera/_utils/type_util.py

```diff
@@ -0,0 +1,93 @@
+"""Module that handles types and annotations."""
+
+from typing import Any, Iterable, Optional, Tuple, Type, TypeVar, Union, cast, overload
+
+from typing_extensions import TypeAlias
+
+try:
+    from types import UnionType  # type: ignore
+except ImportError:
+    UnionType = Union  # type: ignore
+
+try:
+    from typing import Annotated, get_args, get_origin  # type: ignore
+except ImportError:
+    from typing_extensions import Annotated, get_args, get_origin  # type: ignore
+
+
+def is_annotated(annotation: Any):
+    """Check annotation has Annotated type or not."""
+    return get_origin(annotation) is Annotated
+
+
+_Types: TypeAlias = Union[type, Tuple["_Types", ...]]
+
+
+def has_annotated_metadata(annotation: Any, type_: _Types) -> bool:
+    """If given annotation has metadata typed type_, return True."""
+    args = get_args(annotation)
+    for arg in args[1:]:
+        if isinstance(arg, type_):
+            return True
+    return False
+
+
+def consume_annotated_type(annotation: Any) -> type:
+    """If the given type is annotated, return unsubscripted version. If not return itself."""
+    if is_annotated(annotation):
+        return get_args(annotation)[0]
+    return annotation
+
+
+T = TypeVar("T")
+V = TypeVar("V")
+
+
+@overload
+def consume_annotated_metadata(_: Any, __: Type[T]) -> Optional[T]: ...
+
+
+@overload
+def consume_annotated_metadata(_: Any, __: Tuple[Type[T], Type[V]]) -> Optional[Union[T, V]]: ...
+
+
+# FIXME: Currently, mypy cannot guess following type hint properly: https://github.com/python/mypy/issues/17700
+#        def consume_annotated_metadata(_: Any, __: Union[Type[T], Tuple[Type[T], ...]]) -> Optional[T]: ...
+#        Once fixed, remove overloads and add simpler type hints.
+def consume_annotated_metadata(annotation, type_):
+    """If given annotation has metadata typed type_, return the metadata."""
+    args = get_args(annotation)
+    for arg in args[1:]:
+        if isinstance(type_, Iterable):
+            for t in type_:
+                if isinstance(arg, t):
+                    return arg
+        else:
+            if isinstance(arg, type_):
+                return arg
+    return None
+
+
+def may_cast_subscripted_type(t: type) -> type:
+    """If the given type is subscripted, cast and return it.
+
+    This can be helpful if you want to use t with isinstance, issubclass, or etc.,
+    """
+    if origin_type := get_origin(t):
+        return cast(type, origin_type)
+    return t
+
+
+def can_consume_primitive(annotation: type, target: type) -> bool:
+    """If annotation can be considered as target type or subclass of it, return True."""
+    if is_annotated(annotation):
+        annotation = consume_annotated_type(annotation)
+    casted = may_cast_subscripted_type(annotation)
+    if casted is Union or casted is UnionType:
+        return any(can_consume_primitive(arg, target) for arg in get_args(annotation))
+    return issubclass(casted, target)
```

Comment:
We are not consuming anything here. And there are some confusing variable/function names and repeated execution (no need to check `is_annotated` if we check that inside `unwrap_annotation`). We should follow a combination of `get_origin`/`issubclass` naming precedent:

```suggestion
def origin_type_issubclass(cls: type, type_: type) -> bool:
    """Return True if cls can be considered as a subclass of target."""
    unwrapped_type = unwrap_annotation(annotation)
    origin_type = get_origin_or_builtin(unwrapped_type)
    if origin_type is Union or origin_type is UnionType:
        return any(origin_type_issubclass(arg, type_) for arg in get_args(annotation))
    return issubclass(origin_type, type_)
```

Reply:
I agree with naming issue, and thank you for simpler codes!